File: cmd/update_check_test.go
package cmd

import (
	"testing"

	"github.com/spf13/cobra"
)

func TestNormalizeVersion(t *testing.T) {
	cases := map[string]string{
		"v1.2.3":        "1.2.3",
		"1.2.3":         "1.2.3",
		"V0.1.0":        "0.1.0",
		"v1.2.3-beta.1": "1.2.3",
		"  v2.0.0+build ": "2.0.0",
	}
	for in, want := range cases {
		if got := normalizeVersion(in); got != want {
			t.Fatalf("normalizeVersion(%q)=%q, want %q", in, got, want)
		}
	}
}

func TestIsNewerVersion(t *testing.T) {
	newerCases := []struct{ latest, current string }{
		{"0.2.0", "0.1.0"},
		{"v1.0.0", "0.9.9"},
		{"1.2.3", "1.2.2"},
		{"2.0.0", "1.9.9"},
	}
	for _, tc := range newerCases {
		if !isNewerVersion(tc.latest, tc.current) {
			t.Fatalf("expected %q to be newer than %q", tc.latest, tc.current)
		}
	}

	notNewerCases := []struct{ latest, current string }{
		{"0.1.0", "0.1.0"},
		{"0.1.0", "0.2.0"},
		{"v1.2.3", "1.2.3"},
		{"1.2.3", "1.2.4"},
	}
	for _, tc := range notNewerCases {
		if isNewerVersion(tc.latest, tc.current) {
			t.Fatalf("expected %q not to be newer than %q", tc.latest, tc.current)
		}
	}
}

func TestShouldNotifyUpdates_DisabledCases(t *testing.T) {
	oldVersion := Version
	t.Cleanup(func() { Version = oldVersion })

	cmd := &cobra.Command{Use: "test"}
	cmd.Flags().BoolVar(&agentModeFlag, "agent-mode", false, "")

	Version = "dev"
	if shouldNotifyUpdates(cmd) {
		t.Fatal("expected dev version to disable update notices")
	}

	Version = "0.1.0"
	t.Setenv(updateCheckEnvDisable, "1")
	if shouldNotifyUpdates(cmd) {
		t.Fatal("expected VANTA_NO_UPDATE to disable update notices")
	}
	t.Setenv(updateCheckEnvDisable, "")

	t.Setenv("CI", "true")
	if shouldNotifyUpdates(cmd) {
		t.Fatal("expected CI=true to disable update notices")
	}
	t.Setenv("CI", "")

	t.Setenv("CURSOR_AGENT", "1")
	if shouldNotifyUpdates(cmd) {
		t.Fatal("expected agent mode to disable update notices")
	}
}

func TestDisplayVersion(t *testing.T) {
	if got := displayVersion("0.1.0"); got != "v0.1.0" {
		t.Fatalf("displayVersion(0.1.0)=%q", got)
	}
	if got := displayVersion("v0.1.0"); got != "v0.1.0" {
		t.Fatalf("displayVersion(v0.1.0)=%q", got)
	}
	if got := displayVersion("dev"); got != "dev" {
		t.Fatalf("displayVersion(dev)=%q", got)
	}
}

File: src/update-check.test.ts
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import {
  envTruthy,
  isNewerVersion,
  resetUpdateCheckStateForTests,
  shouldNotifyUpdates,
} from "./update-check.js";
import { displayVersion, normalizeVersion } from "./version.js";

describe("normalizeVersion", () => {
  it("strips v prefix and pre-release metadata", () => {
    assert.equal(normalizeVersion("v1.2.3"), "1.2.3");
    assert.equal(normalizeVersion("1.2.3"), "1.2.3");
    assert.equal(normalizeVersion("V0.1.0"), "0.1.0");
    assert.equal(normalizeVersion("v1.2.3-beta.1"), "1.2.3");
    assert.equal(normalizeVersion("  v2.0.0+build "), "2.0.0");
  });
});

describe("displayVersion", () => {
  it("formats versions for display", () => {
    assert.equal(displayVersion("0.1.0"), "v0.1.0");
    assert.equal(displayVersion("v0.1.0"), "v0.1.0");
    assert.equal(displayVersion("dev"), "dev");
  });
});

describe("isNewerVersion", () => {
  it("detects newer semver", () => {
    assert.equal(isNewerVersion("0.2.0", "0.1.0"), true);
    assert.equal(isNewerVersion("v1.0.0", "0.9.9"), true);
    assert.equal(isNewerVersion("1.2.3", "1.2.2"), true);
    assert.equal(isNewerVersion("2.0.0", "1.9.9"), true);
  });

  it("rejects equal or older versions", () => {
    assert.equal(isNewerVersion("0.1.0", "0.1.0"), false);
    assert.equal(isNewerVersion("0.1.0", "0.2.0"), false);
    assert.equal(isNewerVersion("v1.2.3", "1.2.3"), false);
    assert.equal(isNewerVersion("1.2.3", "1.2.4"), false);
  });
});

describe("shouldNotifyUpdates", () => {
  it("disables notices in common non-interactive cases", () => {
    resetUpdateCheckStateForTests();
    // Version is "dev" in tests unless compiled with a define.
    assert.equal(
      shouldNotifyUpdates({ stderrIsTTY: true, env: {} }),
      false,
      "dev version should disable",
    );

    assert.equal(envTruthy("1"), true);
    assert.equal(envTruthy("true"), true);
    assert.equal(envTruthy("0"), false);
  });
});
